tests/test_registrable.py

```diff
@@ -1,27 +1,191 @@
-from pie_core.registrable import Registrable
+from typing import TypeVar
+
+import pytest
+
+from pie_core.registrable import Registrable, RegistrableProtocol, RegistrationError
+
+
+class A(Registrable):
+    """A base class."""
+
+    pass
+
+
+@A.register()
+class B(A):
+    """A registered subclass."""
+
+    pass
+
+
+class C(A):
+    """Unregistered subclass."""
+
+    pass
+
+
+class D(Registrable):
+    """Unregistered base class with no children."""
+
+    pass
 
 
 def test_registrable():
-    class A(Registrable):
+    """A basic Registrable class usage."""
+    # Check the registry entries are created
+    assert A in Registrable._registry
+    assert Registrable._registry[A] == {"B": B}
+
+    # Instantiate an Object of type B by its name.
+    clazz = A.by_name("B")()
+    assert isinstance(clazz, B)
+
+
+def test_base_class():
+
+    assert A.base_class() is A
+    assert B.base_class() is A
+    assert C.base_class() is A
+
+    with pytest.raises(RegistrationError) as e:
+        D.base_class()
+    assert str(e.value) == (
+        f"{D.__name__} has no defined base class. "
+        f"Please annotate {D.__name__} with @<SOME-PARENT-OF-{D.__name__}>.register() "
+        f"to register it at <SOME-PARENT-OF-{D.__name__}>."
+    )
+
+
+def test_has_base_class():
+    assert A.has_base_class()
+    assert B.has_base_class()
+    assert C.has_base_class()
+    assert not D.has_base_class()
+
+
+def test_register():
+    class Base(Registrable):
         pass
 
-    @A.register()
-    class B(A):
+    class Test(Base):
         pass
 
-    assert A in Registrable._registry
-    assert Registrable._registry[A] == {"B": B}
+    # Test main functionality
+    assert not Base.BASE_CLASS
+    assert not Test.BASE_CLASS
```

Comment:
dont use boolean conversion, it is ambiguous (BASE_CLASS may be an empty list in your test), better:
```suggestion
    assert Base.BASE_CLASS is None
    assert Test.BASE_CLASS is None
```